include/eosio/vm/profile.hpp

```diff
@@ -0,0 +1,419 @@
+      EOS_VM_ASSERT(res == 0, profile_exception, "Failed to start timer");
+      struct itimerspec spec;
+      spec.it_interval.tv_sec = 0;
+      spec.it_interval.tv_nsec = profile_interval_us * 1000;
+      spec.it_value.tv_sec = 0;
+      spec.it_value.tv_nsec = profile_interval_us * 1000;
+      res = timer_settime(timer, 0, &spec, nullptr);
+      EOS_VM_ASSERT(res == 0, profile_exception, "Failed to start timer");
+   }
+   void start(profile_data* data) {
+      EOS_VM_ASSERT(!current_data, profile_exception, "Already profiling in the current thread");
+      current_data = data;
+   }
+   void stop() {
+      current_data = nullptr;
+   }
+   ~profile_manager() {
+      current_data = nullptr;
+      timer_delete(timer);
+   }
+   timer_t timer;
+   std::atomic<profile_data*> current_data;
+};
+
+__attribute__((visibility("default")))
+inline thread_local std::unique_ptr<profile_manager> per_thread_profile_manager;
+
+struct scoped_profile {
+   explicit scoped_profile(profile_data* data) {
+      if(data) {
+         if(!per_thread_profile_manager) {
+            per_thread_profile_manager = std::make_unique<profile_manager>();
+         }
+         per_thread_profile_manager->start(data);
+      }
+   }
+   ~scoped_profile() {
+      if(per_thread_profile_manager) {
+         per_thread_profile_manager->stop();
+      }
+   }
+};
+
+#else
+
+__attribute__((visibility("default")))
+inline thread_local std::atomic<profile_data*> per_thread_profile_data = ATOMIC_VAR_INIT(nullptr);
+
+inline void profile_handler(int sig, siginfo_t* info, void* uc) {
+   static_assert(std::atomic<profile_data*>::is_always_lock_free);
+   auto * ptr = std::atomic_load(&per_thread_profile_data);
+   if(ptr) {
+      int saved_errno = errno;
+      void* data[profile_data::max_frames*2]; // Includes both wasm and native frames
+      int count = ptr->get_backtrace_fn(ptr->exec_context, data, sizeof(data)/sizeof(data[0]), uc);
+      ptr->handle_tick(data, count);
+      errno = saved_errno;
+   }
+}
+
+struct profile_manager {
+   profile_manager() {
+      register_profile_signal_handler();
+      timer_thread = std::thread([this]{
+         auto lock = std::unique_lock(mutex);
+         while(!timer_cond.wait_for(lock, std::chrono::microseconds(profile_interval_us), [&]{ return done; })) {
+            for(pthread_t notify : threads_to_notify) {
+               pthread_kill(notify, SIGPROF);
+            }
+         }
+      });
+   }
+   void start(profile_data* data) {
+      auto lock = std::lock_guard(mutex);
+
+      per_thread_profile_data = data;
+      threads_to_notify.push_back(pthread_self());
+   }
+   void stop() {
+      per_thread_profile_data = nullptr;
+      auto lock = std::lock_guard(mutex);
+      threads_to_notify.erase(std::find(threads_to_notify.begin(), threads_to_notify.end(), pthread_self()));
+   }
+   ~profile_manager() {
+      {
+         auto lock = std::unique_lock(mutex);
```

Comment:
Could be a `lock_guard`, not that it matters much. `scoped_lock` even better if eos vm is C++17.